File: platforms/apple/Sources/BotaAppleSDK/Host/HostEffectExecutor.swift
import BotaDeviceSDKC
import Foundation

actor HostEffectExecutor: CoreHost {
    typealias ProgressHandler = @Sendable (UInt64, UInt64) async -> Void

    private let bluetooth: any BluetoothHost
    private let persistence: any PersistenceHost
    private let network: any NetworkHost
    private let material: any MaterialHost
    private let recordingSink: any RecordingSinkHost
    private let firmwareBlob: any FirmwareBlobHost
    private let progress: ProgressHandler
    private var tasks: [CoreCancellationID: [UInt64: Task<Void, Never>]] = [:]
    private var timers: [UInt64: (cancellationID: CoreCancellationID, requestID: UInt64)] = [:]

    init(
        bluetooth: any BluetoothHost,
        persistence: any PersistenceHost,
        network: any NetworkHost,
        material: any MaterialHost,
        recordingSink: any RecordingSinkHost,
        firmwareBlob: any FirmwareBlobHost,
        progress: @escaping ProgressHandler = { _, _ in }
    ) {
        self.bluetooth = bluetooth
        self.persistence = persistence
        self.network = network
        self.material = material
        self.recordingSink = recordingSink
        self.firmwareBlob = firmwareBlob
        self.progress = progress
    }

    func execute(_ effect: CoreEffect) async -> AsyncThrowingStream<CoreHostEvent, Error> {
        switch effect {
        case .timerSchedule:
            return scheduleTimer(effect)
        case .timerCancel:
            return cancelTimer(effect)
        case .persistenceLoadCheckpoint, .persistenceSaveCheckpoint, .persistenceDeleteCheckpoint,
             .persistenceSaveConnectionIdentity, .persistenceSaveFactoryResetResult,
             .persistenceDeleteFactoryResetResult, .secureStorageRead, .secureStorageWrite,
             .secureStorageDelete:
            return route(
                await persistence.execute(effect),
                effect: effect,
                failureKind: UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_PERSISTENCE_FAILED)
            )
        case .bluetoothStartScan, .bluetoothStopScan, .bluetoothConnect,
             .bluetoothDiscoverServices, .bluetoothDisconnect, .bluetoothRead,
             .bluetoothWrite, .bluetoothSubscribe, .bluetoothUnsubscribe:
            return route(
                await bluetooth.execute(effect),
                effect: effect,
                failureKind: UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_BLE_FAILED)
            )
        case .networkDownload, .networkUpload:
            return route(
                await network.execute(effect),
                effect: effect,
                failureKind: UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_NETWORK_FAILED)
            )
        case .progress:
            return reportProgress(effect)
        case .prepareProvisioning, .prepareFactoryResetGrant:
            return route(
                await material.execute(effect),
                effect: effect,
                failureKind: UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_HOST_MATERIAL_FAILED)
            )
        case .recordingSinkTruncate, .recordingSinkAppend, .recordingSinkFinalize,
             .recordingSinkDiscard:
            return route(
                await recordingSink.execute(effect),
                effect: effect,
                failureKind: UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_RECORDING_SINK_FAILED)
            )
        case .firmwareBlobRead:
            return route(
                await firmwareBlob.execute(effect),
                effect: effect,
                failureKind: UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_FIRMWARE_BLOB_FAILED)
            )
        }
    }

    func cancel(_ cancellationID: CoreCancellationID) async {
        let ownedTasks = tasks.removeValue(forKey: cancellationID)?.values ?? [:].values
        for task in ownedTasks { task.cancel() }
        timers = timers.filter { $0.value.cancellationID != cancellationID }
    }

    private func route(
        _ upstream: AsyncThrowingStream<CoreHostEventPayload, Error>,
        effect: CoreEffect,
        failureKind: UInt32
    ) -> AsyncThrowingStream<CoreHostEvent, Error> {
        let pair = AsyncThrowingStream<CoreHostEvent, Error>.makeStream()
        let task = Task {
            do {
                var eventCount = 0
                for try await payload in upstream {
                    try Task.checkCancellation()
                    eventCount += 1
                    guard expectedEventKinds(for: effect).contains(payload.kind),
                          allowsMultipleEvents(effect) || eventCount == 1
                    else {
                        throw invalidEffect(effect, "host returned an event that does not match the effect")
                    }
                    pair.continuation.yield(CoreHostEvent(effect: effect, payload: payload))
                }
                pair.continuation.finish()
            } catch is CancellationError {
                pair.continuation.finish()
            } catch {
                pair.continuation.yield(failureEvent(effect: effect, kind: failureKind, error: error))
                pair.continuation.finish()
            }
            removeTask(effect)
        }
        register(task, for: effect)
        pair.continuation.onTermination = { @Sendable _ in task.cancel() }
        return pair.stream
    }

    private func scheduleTimer(_ effect: CoreEffect) -> AsyncThrowingStream<CoreHostEvent, Error> {
        guard let timerID = effect.packet.fields.unsigned(UInt32(BOTA_DEVICE_SDK_V1_FIELD_TIMER_ID)),
              let delay = effect.packet.fields.unsigned(UInt32(BOTA_DEVICE_SDK_V1_FIELD_DELAY_MS))
        else {
            return failedStream(invalidEffect(effect, "timer schedule fields are missing"))
        }

        let pair = AsyncThrowingStream<CoreHostEvent, Error>.makeStream()
        let task = Task {
            do {
                let nanoseconds = delay.multipliedReportingOverflow(by: 1_000_000)
                try await Task.sleep(nanoseconds: nanoseconds.overflow ? UInt64.max : nanoseconds.partialValue)
                try Task.checkCancellation()
                pair.continuation.yield(CoreHostEvent(
                    effect: effect,
                    kind: UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_TIMER_FIRED),
                    fields: [.unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_TIMER_ID), value: timerID)]
                ))
                pair.continuation.finish()
            } catch {
                pair.continuation.finish()
            }
            removeTimer(timerID, effect: effect)
        }
        timers[timerID] = (effect.cancellationID, effect.requestID)
        register(task, for: effect)
        pair.continuation.onTermination = { @Sendable _ in task.cancel() }
        return pair.stream
    }

    private func cancelTimer(_ effect: CoreEffect) -> AsyncThrowingStream<CoreHostEvent, Error> {
        guard let timerID = effect.packet.fields.unsigned(UInt32(BOTA_DEVICE_SDK_V1_FIELD_TIMER_ID)) else {
            return failedStream(invalidEffect(effect, "timer cancellation ID is missing"))
        }
        if let owner = timers.removeValue(forKey: timerID) {
            tasks[owner.cancellationID]?[owner.requestID]?.cancel()
            tasks[owner.cancellationID]?[owner.requestID] = nil
        }
        return emptyStream()
    }

    private func reportProgress(_ effect: CoreEffect) -> AsyncThrowingStream<CoreHostEvent, Error> {
        guard let completed = effect.packet.fields.unsigned(UInt32(BOTA_DEVICE_SDK_V1_FIELD_COMPLETED_UNITS)),
              let total = effect.packet.fields.unsigned(UInt32(BOTA_DEVICE_SDK_V1_FIELD_TOTAL_UNITS))
        else {
            return failedStream(invalidEffect(effect, "progress fields are missing"))
        }
        let pair = AsyncThrowingStream<CoreHostEvent, Error>.makeStream()
        Task {
            await progress(completed, total)
            pair.continuation.finish()
        }
        return pair.stream
    }

    private func failureEvent(effect: CoreEffect, kind: UInt32, error: Error) -> CoreHostEvent {
        let platformCode = Int64((error as NSError).code)
        var fields: [CoreField] = []
        if kind == UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_NETWORK_FAILED) {
            let transferID = effect.packet.fields.unsigned(UInt32(BOTA_DEVICE_SDK_V1_FIELD_DOWNLOAD_ID))
                ?? effect.packet.fields.unsigned(UInt32(BOTA_DEVICE_SDK_V1_FIELD_UPLOAD_ID))
                ?? 0
            fields.append(.unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_TRANSFER_ID), value: transferID))
            if case let NativeHostError.httpStatus(status) = error {
                fields.append(.unsigned(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_STATUS_CODE), value: UInt64(status)))
            }
        } else {
            fields.append(.signed(id: UInt32(BOTA_DEVICE_SDK_V1_FIELD_PLATFORM_CODE), value: platformCode))
        }
        return CoreHostEvent(effect: effect, kind: kind, fields: fields)
    }

    private func expectedEventKinds(for effect: CoreEffect) -> Set<UInt32> {
        switch effect {
        case .timerSchedule:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_TIMER_FIRED)]
        case .timerCancel, .progress, .bluetoothUnsubscribe, .recordingSinkDiscard:
            return []
        case .persistenceLoadCheckpoint:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_CHECKPOINT_LOADED)]
        case .persistenceSaveCheckpoint, .persistenceDeleteCheckpoint:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_CHECKPOINT_SAVED)]
        case .persistenceSaveConnectionIdentity:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_CONNECTION_IDENTITY_SAVED)]
        case .persistenceSaveFactoryResetResult:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_FACTORY_RESET_RESULT_SAVED)]
        case .persistenceDeleteFactoryResetResult:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_FACTORY_RESET_RESULT_DELETED)]
        case .secureStorageRead:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_SECRET_LOADED)]
        case .secureStorageWrite, .secureStorageDelete:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_SECRET_STORED)]
        case .bluetoothStartScan:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_BLE_SCAN_RESULT)]
        case .bluetoothStopScan:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_BLE_SCAN_STOPPED)]
        case .bluetoothConnect:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_BLE_CONNECTED)]
        case .bluetoothDiscoverServices:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_BLE_SERVICES_DISCOVERED)]
        case .bluetoothDisconnect:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_BLE_DISCONNECTED)]
        case .bluetoothRead:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_BLE_READ_COMPLETED)]
        case .bluetoothWrite:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_BLE_WRITE_COMPLETED)]
        case .bluetoothSubscribe:
            return [
                UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_BLE_SUBSCRIBED),
                UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_BLE_NOTIFICATION),
                UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_BLE_DISCONNECTED),
            ]
        case .networkDownload:
            return [
                UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_NETWORK_DOWNLOAD_PROGRESS),
                UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_NETWORK_DOWNLOAD_COMPLETED),
            ]
        case .networkUpload:
            return [
                UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_NETWORK_UPLOAD_PROGRESS),
                UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_NETWORK_UPLOAD_COMPLETED),
            ]
        case .prepareProvisioning:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_PROVISIONING_MATERIAL_PREPARED)]
        case .prepareFactoryResetGrant:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_FACTORY_RESET_GRANT_PREPARED)]
        case .recordingSinkTruncate:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_RECORDING_SINK_TRUNCATED)]
        case .recordingSinkAppend:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_RECORDING_SINK_APPEND_COMPLETED)]
        case .recordingSinkFinalize:
            return [
                UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_RECORDING_SINK_FINALIZED),
                UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_RECORDING_SINK_INTEGRITY_FAILED),
            ]
        case .firmwareBlobRead:
            return [UInt32(BOTA_DEVICE_SDK_V1_HOST_EVENT_FIRMWARE_CHUNK_READ)]
        }
    }

    private func allowsMultipleEvents(_ effect: CoreEffect) -> Bool {
        switch effect {
        case .bluetoothStartScan, .bluetoothSubscribe, .networkDownload, .networkUpload:
            return true
        case .timerSchedule, .timerCancel, .persistenceLoadCheckpoint,
             .persistenceSaveCheckpoint, .persistenceDeleteCheckpoint,
             .persistenceSaveConnectionIdentity, .persistenceSaveFactoryResetResult,
             .persistenceDeleteFactoryResetResult, .secureStorageRead, .secureStorageWrite,
             .secureStorageDelete, .bluetoothStopScan, .bluetoothConnect,
             .bluetoothDiscoverServices, .bluetoothDisconnect, .bluetoothRead,
             .bluetoothWrite, .bluetoothUnsubscribe, .progress, .prepareProvisioning,
             .prepareFactoryResetGrant, .recordingSinkTruncate, .recordingSinkAppend,
             .recordingSinkFinalize, .recordingSinkDiscard, .firmwareBlobRead:
            return false
        }
    }

    private func register(_ task: Task<Void, Never>, for effect: CoreEffect) {
        tasks[effect.cancellationID, default: [:]][effect.requestID] = task
    }

    private func removeTask(_ effect: CoreEffect) {
        tasks[effect.cancellationID]?[effect.requestID] = nil
        if tasks[effect.cancellationID]?.isEmpty == true { tasks[effect.cancellationID] = nil }
    }

    private func removeTimer(_ timerID: UInt64, effect: CoreEffect) {
        if timers[timerID]?.requestID == effect.requestID { timers[timerID] = nil }
        removeTask(effect)
    }

    private func emptyStream() -> AsyncThrowingStream<CoreHostEvent, Error> {
        AsyncThrowingStream { $0.finish() }
    }

    private func failedStream(_ error: Error) -> AsyncThrowingStream<CoreHostEvent, Error> {
        AsyncThrowingStream { $0.finish(throwing: error) }
    }

    private func invalidEffect(_ effect: CoreEffect, _ detail: String) -> CoreError {
        CoreError(
            code: UInt32(BOTA_DEVICE_SDK_V1_ERROR_INVALID_INPUT),
            operation: effect.operation,
            retryable: false,
            protocolStatus: nil,
            detail: detail
        )
    }
}

File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/internal/host/HostEffectExecutor.kt
package dev.bota.sdk.internal.host

import dev.bota.sdk.BotaErrorCode
import dev.bota.sdk.BotaSDKError
import dev.bota.sdk.internal.bluetooth.BluetoothTransportException
import dev.bota.sdk.internal.core.CoreCancellationId
import dev.bota.sdk.internal.core.CoreEffect
import dev.bota.sdk.internal.core.CoreEffectHandler
import dev.bota.sdk.internal.core.CoreEffectKind
import dev.bota.sdk.internal.core.CoreField
import dev.bota.sdk.internal.core.CoreHostEvent
import dev.bota.sdk.internal.core.HostEventKind
import kotlin.math.absoluteValue
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.emptyFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.launch

internal class HostEffectExecutor(
    private val bluetooth: BluetoothHost,
    private val persistence: PersistenceHost,
    private val secureStorage: SecureStorageHost,
    private val network: NetworkHost,
    private val material: MaterialHost,
    private val recordingSink: RecordingSinkHost,
    private val firmwareBlob: FirmwareBlobHost,
    private val encryptedUploadV2: EncryptedUploadV2Host = UnavailableEncryptedUploadV2Host(),
    private val progress: suspend (completed: ULong, total: ULong) -> Unit = { _, _ -> },
) : CoreEffectHandler {
    private data class TimerOwner(
        val cancellationId: CoreCancellationId,
        val requestId: ULong,
        val job: Job,
    )

    private val timerLock = Any()
    private val timers = mutableMapOf<ULong, TimerOwner>()

    override fun execute(effect: CoreEffect): Flow<CoreHostEvent> = when (effect.kind) {
        CoreEffectKind.TimerSchedule -> scheduleTimer(effect)
        CoreEffectKind.TimerCancel -> cancelTimer(effect)
        CoreEffectKind.PersistenceLoadCheckpoint,
        CoreEffectKind.PersistenceSaveCheckpoint,
        CoreEffectKind.PersistenceDeleteCheckpoint,
        CoreEffectKind.PersistenceSaveConnectionIdentity,
        CoreEffectKind.PersistenceSaveFactoryResetResult,
        CoreEffectKind.PersistenceDeleteFactoryResetResult ->
            route(effect, persistence.execute(effect), HostEventKind.PersistenceFailed)
        CoreEffectKind.SecureStorageRead,
        CoreEffectKind.SecureStorageWrite,
        CoreEffectKind.SecureStorageDelete ->
            route(effect, secureStorage.execute(effect), HostEventKind.PersistenceFailed)
        CoreEffectKind.BluetoothStartScan,
        CoreEffectKind.BluetoothStopScan,
        CoreEffectKind.BluetoothConnect,
        CoreEffectKind.BluetoothDiscoverServices,
        CoreEffectKind.BluetoothDisconnect,
        CoreEffectKind.BluetoothRead,
        CoreEffectKind.BluetoothWrite,
        CoreEffectKind.BluetoothSubscribe,
        CoreEffectKind.BluetoothUnsubscribe ->
            route(effect, bluetooth.execute(effect), HostEventKind.BleFailed)
        CoreEffectKind.NetworkDownload,
        CoreEffectKind.NetworkUpload ->
            route(effect, network.execute(effect), HostEventKind.NetworkFailed)
        CoreEffectKind.Progress -> reportProgress(effect)
        CoreEffectKind.PrepareProvisioning,
        CoreEffectKind.PrepareFactoryResetGrant ->
            route(effect, material.execute(effect), HostEventKind.HostMaterialFailed)
        CoreEffectKind.RecordingSinkTruncate,
        CoreEffectKind.RecordingSinkAppend,
        CoreEffectKind.RecordingSinkFinalize,
        CoreEffectKind.RecordingSinkDiscard,
        CoreEffectKind.StreamingSinkAppendPlaintext,
        CoreEffectKind.StreamingSinkBeginEncrypted,
        CoreEffectKind.StreamingSinkAppendEncrypted,
        CoreEffectKind.StreamingSinkFinalize,
        CoreEffectKind.StreamingSinkDiscard ->
            route(
                effect,
                recordingSink.execute(effect),
                if (effect.kind.isStreamingSink) HostEventKind.StreamingSinkFailed
                else HostEventKind.RecordingSinkFailed,
            )
        CoreEffectKind.FirmwareBlobRead ->
            route(effect, firmwareBlob.execute(effect), HostEventKind.FirmwareBlobFailed)
        CoreEffectKind.EncryptedUploadV2LoadCheckpoint,
        CoreEffectKind.EncryptedUploadV2DeleteCheckpoint,
        CoreEffectKind.EncryptedUploadV2TruncateSink,
        CoreEffectKind.EncryptedUploadV2PrepareSession,
        CoreEffectKind.EncryptedUploadV2StartTransfer,
        CoreEffectKind.EncryptedUploadV2RepairWindow,
        CoreEffectKind.EncryptedUploadV2SaveCheckpoint,
        CoreEffectKind.EncryptedUploadV2AcknowledgeWindow,
        CoreEffectKind.EncryptedUploadV2StageArtifacts,
        CoreEffectKind.EncryptedUploadV2AwaitReceipt,
        CoreEffectKind.EncryptedUploadV2ConfirmWithReceipt,
        CoreEffectKind.EncryptedUploadV2Abort ->
            route(effect, encryptedUploadV2.execute(effect), HostEventKind.EncryptedUploadV2Failed)
    }

    override suspend fun cancel(cancellationId: CoreCancellationId) {
        val jobs = synchronized(timerLock) {
            val owned = timers.filterValues { it.cancellationId == cancellationId }.values.map(TimerOwner::job)
            timers.entries.removeAll { it.value.cancellationId == cancellationId }
            owned
        }
        jobs.forEach(Job::cancel)
        encryptedUploadV2.cancel(cancellationId)
    }

    private fun route(
        effect: CoreEffect,
        upstream: Flow<CoreHostEventPayload>,
        failureKind: HostEventKind,
    ): Flow<CoreHostEvent> = flow {
        var eventCount = 0
        try {
            upstream.collect { payload ->
                eventCount += 1
                if (payload.kind !in expectedEventKinds(effect.kind) ||
                    (!allowsMultipleEvents(effect.kind) && eventCount > 1)
                ) {
                    throw NativeHostException(1, "host returned an event that does not match ${effect.kind}")
                }
                if (payload.fields.rawByteCount() > CoreEffect.MaximumRawByteCount) {
                    throw NativeHostException(4, "host event contains too many raw bytes")
                }
                emit(CoreHostEvent.fromEffect(effect, payload.kind, payload.fields))
            }
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            emit(failureEvent(effect, failureKind, error))
        }
    }

    private fun scheduleTimer(effect: CoreEffect): Flow<CoreHostEvent> {
        val timerId = effect.packet.unsigneds(26).firstOrNull()
            ?: return failedFlow("timer schedule ID is missing")
        val delayMs = effect.packet.unsigneds(27).firstOrNull()
            ?: return failedFlow("timer schedule delay is missing")
        return callbackFlow {
            val timerJob = launch {
                try {
                    delay(delayMs.coerceAtMost(Long.MAX_VALUE.toULong()).toLong())
                    send(
                        CoreHostEvent.fromEffect(
                            effect,
                            HostEventKind.TimerFired,
                            listOf(CoreField.Unsigned(26, timerId)),
                        ),
                    )
                } finally {
                    close()
                }
            }
            synchronized(timerLock) {
                timers.remove(timerId)?.job?.cancel()
                timers[timerId] = TimerOwner(effect.cancellationId, effect.requestId, timerJob)
            }
            awaitClose {
                timerJob.cancel()
                synchronized(timerLock) {
                    if (timers[timerId]?.requestId == effect.requestId) timers.remove(timerId)
                }
            }
        }
    }

    private fun cancelTimer(effect: CoreEffect): Flow<CoreHostEvent> {
        val timerId = effect.packet.unsigneds(26).firstOrNull()
            ?: return failedFlow("timer cancellation ID is missing")
        synchronized(timerLock) { timers.remove(timerId) }?.job?.cancel()
        return emptyFlow()
    }

    private fun reportProgress(effect: CoreEffect): Flow<CoreHostEvent> {
        val completed = effect.packet.unsigneds(36).firstOrNull()
            ?: return failedFlow("progress completed units are missing")
        val total = effect.packet.unsigneds(15).firstOrNull()
            ?: return failedFlow("progress total units are missing")
        return flow { progress(completed, total) }
    }

    private fun failureEvent(
        effect: CoreEffect,
        kind: HostEventKind,
        error: Throwable,
    ): CoreHostEvent {
        val hostError = error as? NativeHostException
        val fields = if (kind == HostEventKind.NetworkFailed) {
            buildList<CoreField> {
                val transferId = effect.packet.unsigneds(59).firstOrNull()
                    ?: effect.packet.unsigneds(21).firstOrNull()
                    ?: 0u
                add(CoreField.Unsigned(59, transferId))
                hostError?.httpStatus?.let { add(CoreField.Unsigned(60, it.toULong())) }
            }
        } else if (kind == HostEventKind.EncryptedUploadV2Failed) {
            val hostError = encryptedUploadV2Failure(error)
            buildList<CoreField> {
                add(CoreField.Unsigned(47, hostError.errorCode.toULong()))
                add(CoreField.BooleanValue(48, hostError.retryable))
                hostError.protocolStatus?.let { add(CoreField.Unsigned(49, it.toULong())) }
                error.message?.let { add(CoreField.Text(50, it)) }
            }
        } else {
            val platformCode = -(hostError?.platformCode ?: 1).toLong().absoluteValue
            listOf(CoreField.Signed(52, platformCode))
        }
        return CoreHostEvent.fromEffect(effect, kind, fields)
    }

    private fun failedFlow(detail: String): Flow<CoreHostEvent> = flow {
        throw NativeHostException(1, detail)
    }

    private fun encryptedUploadV2Failure(error: Throwable): EncryptedUploadV2HostException = when (error) {
        is EncryptedUploadV2HostException -> error
        is BotaSDKError.Core -> EncryptedUploadV2HostException(
            error.code.encryptedUploadV2Code(), error.retryable, error.protocolStatus, error.detail,
        )
        is BluetoothTransportException -> EncryptedUploadV2HostException(
            12u, true, message = error.message ?: "Bluetooth transport failed",
        )
        is IllegalArgumentException -> EncryptedUploadV2HostException(
            1u, false, message = error.message ?: "invalid encrypted upload v2 input",
        )
        else -> error.cause?.takeIf { it !== error }?.let(::encryptedUploadV2Failure)
            ?: EncryptedUploadV2HostException(
                21u, false, message = error.message ?: "encrypted upload v2 host failed",
            )
    }
}

private fun BotaErrorCode.encryptedUploadV2Code(): UInt = when (this) {
    BotaErrorCode.InvalidInput -> 1u
    BotaErrorCode.TruncatedPacket -> 2u
    BotaErrorCode.UnknownPacket -> 3u
    BotaErrorCode.PayloadTooLarge -> 4u
    BotaErrorCode.UnsupportedCapability -> 5u
    BotaErrorCode.UnsupportedOperation -> 6u
    BotaErrorCode.FeatureUnavailable -> 7u
    BotaErrorCode.OperationInProgress -> 8u
    BotaErrorCode.UnexpectedEvent -> 9u
    BotaErrorCode.DeviceNotFound -> 10u
    BotaErrorCode.IdentityMismatch -> 11u
    BotaErrorCode.ConnectionFailed -> 12u
    BotaErrorCode.PersistenceFailed -> 13u
    BotaErrorCode.NotConnected -> 14u
    BotaErrorCode.Timeout -> 15u
    BotaErrorCode.Cancelled -> 16u
    BotaErrorCode.ProtocolRejected -> 17u
    BotaErrorCode.IntegrityFailed -> 18u
    BotaErrorCode.UploadOwnershipUnknown -> 19u
    BotaErrorCode.DownloadFailed -> 20u
    BotaErrorCode.Internal -> 21u
    is BotaErrorCode.Unknown -> 21u
}

private fun List<CoreField>.rawByteCount(): Int = sumOf { field ->
    when (field) {
        is CoreField.Text -> field.value.encodeToByteArray().size
        is CoreField.Bytes -> field.value.size
        is CoreField.Unsigned, is CoreField.Signed, is CoreField.BooleanValue -> 0
    }
}

private fun allowsMultipleEvents(kind: CoreEffectKind): Boolean = when (kind) {
    CoreEffectKind.BluetoothStartScan,
    CoreEffectKind.BluetoothSubscribe,
    CoreEffectKind.NetworkDownload,
    CoreEffectKind.NetworkUpload -> true
    CoreEffectKind.TimerSchedule,
    CoreEffectKind.TimerCancel,
    CoreEffectKind.PersistenceLoadCheckpoint,
    CoreEffectKind.PersistenceSaveCheckpoint,
    CoreEffectKind.PersistenceDeleteCheckpoint,
    CoreEffectKind.PersistenceSaveConnectionIdentity,
    CoreEffectKind.PersistenceSaveFactoryResetResult,
    CoreEffectKind.PersistenceDeleteFactoryResetResult,
    CoreEffectKind.SecureStorageRead,
    CoreEffectKind.SecureStorageWrite,
    CoreEffectKind.SecureStorageDelete,
    CoreEffectKind.BluetoothStopScan,
    CoreEffectKind.BluetoothConnect,
    CoreEffectKind.BluetoothDiscoverServices,
    CoreEffectKind.BluetoothDisconnect,
    CoreEffectKind.BluetoothRead,
    CoreEffectKind.BluetoothWrite,
    CoreEffectKind.BluetoothUnsubscribe,
    CoreEffectKind.Progress,
    CoreEffectKind.PrepareProvisioning,
    CoreEffectKind.PrepareFactoryResetGrant,
    CoreEffectKind.RecordingSinkTruncate,
    CoreEffectKind.RecordingSinkAppend,
    CoreEffectKind.RecordingSinkFinalize,
    CoreEffectKind.RecordingSinkDiscard,
    CoreEffectKind.StreamingSinkAppendPlaintext,
    CoreEffectKind.StreamingSinkBeginEncrypted,
    CoreEffectKind.StreamingSinkAppendEncrypted,
    CoreEffectKind.StreamingSinkFinalize,
    CoreEffectKind.StreamingSinkDiscard,
    CoreEffectKind.FirmwareBlobRead -> false
    CoreEffectKind.EncryptedUploadV2LoadCheckpoint,
    CoreEffectKind.EncryptedUploadV2DeleteCheckpoint,
    CoreEffectKind.EncryptedUploadV2TruncateSink,
    CoreEffectKind.EncryptedUploadV2PrepareSession,
    CoreEffectKind.EncryptedUploadV2RepairWindow,
    CoreEffectKind.EncryptedUploadV2SaveCheckpoint,
    CoreEffectKind.EncryptedUploadV2AcknowledgeWindow,
    CoreEffectKind.EncryptedUploadV2StageArtifacts,
    CoreEffectKind.EncryptedUploadV2AwaitReceipt,
    CoreEffectKind.EncryptedUploadV2ConfirmWithReceipt,
    CoreEffectKind.EncryptedUploadV2Abort -> false
    CoreEffectKind.EncryptedUploadV2StartTransfer -> true
}

private fun expectedEventKinds(kind: CoreEffectKind): Set<HostEventKind> = when (kind) {
    CoreEffectKind.TimerSchedule -> setOf(HostEventKind.TimerFired)
    CoreEffectKind.TimerCancel,
    CoreEffectKind.Progress,
    CoreEffectKind.BluetoothUnsubscribe,
    CoreEffectKind.RecordingSinkDiscard -> emptySet()
    CoreEffectKind.PersistenceLoadCheckpoint -> setOf(HostEventKind.CheckpointLoaded)
    CoreEffectKind.PersistenceSaveCheckpoint,
    CoreEffectKind.PersistenceDeleteCheckpoint -> setOf(HostEventKind.CheckpointSaved)
    CoreEffectKind.PersistenceSaveConnectionIdentity -> setOf(HostEventKind.ConnectionIdentitySaved)
    CoreEffectKind.PersistenceSaveFactoryResetResult -> setOf(HostEventKind.FactoryResetResultSaved)
    CoreEffectKind.PersistenceDeleteFactoryResetResult -> setOf(HostEventKind.FactoryResetResultDeleted)
    CoreEffectKind.SecureStorageRead -> setOf(HostEventKind.SecretLoaded)
    CoreEffectKind.SecureStorageWrite,
    CoreEffectKind.SecureStorageDelete -> setOf(HostEventKind.SecretStored)
    CoreEffectKind.BluetoothStartScan -> setOf(HostEventKind.BleScanResult)
    CoreEffectKind.BluetoothStopScan -> setOf(HostEventKind.BleScanStopped)
    CoreEffectKind.BluetoothConnect -> setOf(HostEventKind.BleConnected)
    CoreEffectKind.BluetoothDiscoverServices -> setOf(HostEventKind.BleServicesDiscovered)
    CoreEffectKind.BluetoothDisconnect -> setOf(HostEventKind.BleDisconnected)
    CoreEffectKind.BluetoothRead -> setOf(HostEventKind.BleReadCompleted)
    CoreEffectKind.BluetoothWrite -> setOf(HostEventKind.BleWriteCompleted)
    CoreEffectKind.BluetoothSubscribe -> setOf(
        HostEventKind.BleSubscribed,
        HostEventKind.BleNotification,
        HostEventKind.BleDisconnected,
    )
    CoreEffectKind.NetworkDownload -> setOf(
        HostEventKind.NetworkDownloadProgress,
        HostEventKind.NetworkDownloadCompleted,
    )
    CoreEffectKind.NetworkUpload -> setOf(
        HostEventKind.NetworkUploadProgress,
        HostEventKind.NetworkUploadCompleted,
    )
    CoreEffectKind.PrepareProvisioning -> setOf(HostEventKind.ProvisioningMaterialPrepared)
    CoreEffectKind.PrepareFactoryResetGrant -> setOf(HostEventKind.FactoryResetGrantPrepared)
    CoreEffectKind.RecordingSinkTruncate -> setOf(HostEventKind.RecordingSinkTruncated)
    CoreEffectKind.RecordingSinkAppend -> setOf(HostEventKind.RecordingSinkAppendCompleted)
    CoreEffectKind.RecordingSinkFinalize -> setOf(
        HostEventKind.RecordingSinkFinalized,
        HostEventKind.RecordingSinkIntegrityFailed,
    )
    CoreEffectKind.StreamingSinkAppendPlaintext,
    CoreEffectKind.StreamingSinkBeginEncrypted,
    CoreEffectKind.StreamingSinkAppendEncrypted -> setOf(HostEventKind.StreamingSinkAccepted)
    CoreEffectKind.StreamingSinkFinalize -> setOf(HostEventKind.StreamingSinkFinalized)
    CoreEffectKind.StreamingSinkDiscard -> emptySet()
    CoreEffectKind.FirmwareBlobRead -> setOf(HostEventKind.FirmwareChunkRead)
    CoreEffectKind.EncryptedUploadV2LoadCheckpoint ->
        setOf(HostEventKind.EncryptedUploadV2CheckpointLoaded)
    CoreEffectKind.EncryptedUploadV2DeleteCheckpoint,
    CoreEffectKind.EncryptedUploadV2Abort -> emptySet()
    CoreEffectKind.EncryptedUploadV2TruncateSink ->
        setOf(HostEventKind.EncryptedUploadV2SinkTruncated)
    CoreEffectKind.EncryptedUploadV2PrepareSession ->
        setOf(HostEventKind.EncryptedUploadV2SessionPrepared)
    CoreEffectKind.EncryptedUploadV2StartTransfer -> setOf(
        HostEventKind.EncryptedUploadV2TransferStarted,
        HostEventKind.EncryptedUploadV2ResumeRejected,
        HostEventKind.EncryptedUploadV2WindowStaged,
        HostEventKind.EncryptedUploadV2TransferCompleted,
        HostEventKind.EncryptedUploadV2MixedProfile,
    )
    CoreEffectKind.EncryptedUploadV2RepairWindow ->
        setOf(HostEventKind.EncryptedUploadV2WindowStaged)
    CoreEffectKind.EncryptedUploadV2SaveCheckpoint ->
        setOf(HostEventKind.EncryptedUploadV2CheckpointSaved)
    CoreEffectKind.EncryptedUploadV2AcknowledgeWindow ->
        setOf(HostEventKind.EncryptedUploadV2WindowAcknowledged)
    CoreEffectKind.EncryptedUploadV2StageArtifacts ->
        setOf(HostEventKind.EncryptedUploadV2ArtifactsStaged)
    CoreEffectKind.EncryptedUploadV2AwaitReceipt ->
        setOf(HostEventKind.EncryptedUploadV2ReceiptAccepted)
    CoreEffectKind.EncryptedUploadV2ConfirmWithReceipt ->
        setOf(HostEventKind.EncryptedUploadV2RecordingConfirmed)
}

private val CoreEffectKind.isStreamingSink: Boolean
    get() = this == CoreEffectKind.StreamingSinkAppendPlaintext ||
        this == CoreEffectKind.StreamingSinkBeginEncrypted ||
        this == CoreEffectKind.StreamingSinkAppendEncrypted ||
        this == CoreEffectKind.StreamingSinkFinalize ||
        this == CoreEffectKind.StreamingSinkDiscard
